src/runtime/propensity-server.js

```diff
@@ -0,0 +1,262 @@
+      };
+      xhr.onerror = () => {
+        reject(new Error('Network failure'));
+      };
+      xhr.onabort = () => {
+        reject(new Error('Request aborted'));
+      };
+      if (init.method == 'POST') {
+        xhr.send(init.body);
+      } else {
+        xhr.send();
+      }
+    });
+  }
+
+  /**
+   * Creates header from XHR header
+   * @param {?string} responseHeaders
+   * @return {!Headers}
+   */
+  getResponseHeader_(responseHeaders) {
+    const headerMap = new Headers();
+    if (!responseHeaders) {
+      return headerMap;
+    }
+    const arr = responseHeaders.trim().split(/[/r/n]+/);
+    arr.forEach(function(line) {
+      const parts = line.split(': ');
+      const header = parts.shift();
+      const value = parts.join(': ');
+      headerMap[header] = value;
+    });
+    return headerMap;
+  }
+}
+
+/**
+ * Implements interface to Propensity server
+ */
+export class PropensityServer {
+  /**
+   * Page configuration is known when Propensity API
+   * is available, publication ID is therefore used
+   * in constructor for the server interface.
+   * @param {string} publicationId
+   */
+  constructor(win, publicationId) {
+    /** @private @const {!Window} */
+    this.win_ = win;
+    /** @private @const {string} */
+    this.publicationId_ = publicationId;
+    /** @private {?string} */
+    this.clientId_ = null;
+    /** @private {?string} */
+    this.product_ = null;
+    /** @private {boolean} */
+    this.userConsent_ = false;
+    /** @private @const {boolean}*/
+    this.TEST_SERVERS_ = true;
+    /** {!Object.<number, string>} */
+    const errorMap = {
+      404: 'Publisher not whitelisted',
+      403: 'Not sent from allowed origin',
+      400: 'Invalid request',
+      500: 'Server not available',
+    };
+    /** @private @const {!XhrInterface} */
+    this.xhrInterface_ = new XhrInterface(errorMap);
+  }
+
+  /**
+   * @return {string}
+   */
+  getUrl_() {
+    if (this.TEST_SERVERS_) {
+      return 'http://sohanirao.mtv.corp.google.com:8080';
+    } else {
+      return 'https://pubads.g.doubleclick.net/gampad/adx';
+    }
+  }
+
+  /**
+   * @return {?string}
+   */
+  getGads_() {
+    let gads = null;
+    const gadsmatch = this.win_.document.cookie.match(
+        '(^|;)\\s*__gads\\s*=\\s*([^;]+)');
+    gads = gadsmatch && encodeURIComponent(gadsmatch.pop());
+    return gads;
+  }
+
+  /**
+   * Returns the client ID to be used
+   * @return {?string}
+   * @private
+   */
+  getClientId_() {
+    if (!this.clientId_) {
+      if (!this.userConsent_) {
+        this.clientId_ = 'noConsent';
+      } else {
+        this.clientId_ = this.getGads_();
+      }
+    }
+    return this.clientId_;
+  }
+
+  /**
+   * @param {boolean} userConsent
+   */
+  setUserConsent(userConsent) {
+    this.userConsent_ = userConsent;
+  }
+
+  /**
+   * @param {string} state
+   * @param {?string} entitlements
+   */
+  sendSubscriptionState(state, entitlements) {
+    this.product_ = entitlements;
```

Comment:
nit: comment here to explain that the DRX server takes a 'product' and so we need to "convert" our entitlements to a product. Don't use my words, you can explain it much better.

Reply:
actually this is not necessary. will remove.